File: src/assembler/tokenizer.nim
# Este modulo é responsavel por carregar programas e processar eles para garantir
# a sintaxe correta de cada instrução, remover coisas desnecessarias como comentarios
# e linhas em brancos, e, fazer a troca dos labels de jumps pelo endereço.

import strutils
# import sintaxes

type 
    TokenType* = enum 
        INSTRUCTION,   # instrução
        ARGUMENT,      # argumento para instrução
        LABEL_DEF,     # definição de label
        LABEL_REF,     # uso de labels
        TYPE,          # .string, .word...
        SECTION,       # .text e .data
        OPEN_PARENTH   # (
        CLOSE_PARENTH  # )
        REGISTER,
        COMMENT,
        STRING,
        NUMBER,
        NEW_LINE,
        VIRGULA

    Token* = ref object 
        tokenType*: TokenType
        value*: string
        line*: int
        column*: int

    Tokenizer* = object
        source*: string  
        position*: int   
        line*: int   
        column*: int 
        tokens*: seq[Token]  

var labels: seq[string]

proc `$`*(token: Token): string =
  result = "Token(" & $token.tokenType & ", '" & token.value & "', line=" & $token.line & ")"

proc newTokenzier*(file: string): Tokenizer =
    result.source = file
    result.position = 0
    result.line = 1
    result.column = 1
    result.tokens = @[]

proc advance(self: var Tokenizer, total: int=1) =
    self.position += total
    self.column += total

proc addToken(self: var Tokenizer, token_t: TokenType, value: string) =
    let t: Token = Token(
        tokenType: token_t,
        value: value,
        line: self.line,
        column: self.column
    )

    self.tokens.add(t)

proc currentChar(self: Tokenizer): char = 
    if self.position < self.source.len:
        return self.source[self.position]
    else:
        return '\0' # Fim do arquivo

proc peekNextChar(self: Tokenizer, offset: int=1): char =
    let next: int = self.position + offset
    if next < self.source.len:
        return self.source[next]
    else:
        return '\0'

proc handleWhiteSpace(self: var Tokenizer) = 
    case self.currentChar
        of '\n':
            self.addToken(NEW_LINE, "\n")
            self.advance()
            self.line += 1
            self.column = 1
        else:
            self.advance()

# proc handleComment(self: var Tokenizer) =
#   while self.currentChar != '\n' and self.currentChar != '\0':
#     self.advance()

# TODO - Adicionar suporte a Escape
proc handleString(self: var Tokenizer): bool =
    self.advance()
    var str: string = ""
    while (self.currentChar != '"') :
        let ch: char = self.currentChar()
        if (ch == '\n') :
            echo "Erro ao ler string na linha: ", self.line
            return false
        str.add(ch)
        self.advance()
    self.advance()
    self.addToken(STRING, str)
    return true

proc handleDirective(self: var Tokenizer) =
    let start = self.position  
    self.advance() 
        
    while self.currentChar.isAlphaNumeric() or self.currentChar == '_':
        self.advance() 
        
    let directive: string = self.source.substr(start, self.position - 1)
    var t = TYPE
    if (directive == ".text" or directive == ".data") :
        t = SECTION

    self.addToken(t, directive)  

proc setLastTokenType(self: var Tokenizer, newType: TokenType) =
    self.tokens[^1].tokenType = newType

proc handleLabelDef(self: var Tokenizer) =
    self.setLastTokenType(LABEL_DEF)
    labels.add(self.tokens[^1].value)
    self.advance()

proc handleIdentifier(self: var Tokenizer) =
    let start: int = self.position

    while self.currentChar.isAlphaNumeric() or self.currentChar == '_':
        self.advance()

    let identifier: string = self.source.substr(start, self.position - 1)
    self.addToken(INSTRUCTION, identifier)

proc isHex(c: char): bool =
    if not (('0' <= c and c <= '9') or
            ('a' <= c and c <= 'f') or
            ('A' <= c and c <= 'F')):
        return false
    return true

proc handleNumber(self: var Tokenizer) =
    let start: int = self.position

    # Handle negativo
    if self.currentChar == '-':
        self.advance()

    # Handle hexadecimal
    if self.currentChar == '0' and self.peekNextChar().toLowerAscii() == 'x':
        self.advance(2)  # Pula '0x'
        while self.currentChar().isHex():
            self.advance()
    # Decimal
    else:
        while self.currentChar.isDigit():
            self.advance()

    let number: string = self.source.substr(start, self.position - 1)
    self.addToken(NUMBER, number)

proc handleRegister(self: var Tokenizer) =
    let start: int = self.position
    self.advance() 

    while self.currentChar().isAlphaNumeric() :
        self.advance()

    let register: string = self.source.substr(start, self.position - 1)
    self.addToken(REGISTER, register)

proc handleComment(self: var Tokenizer) =
    let start: int = self.position
    self.advance()

    while self.currentChar() != '\n' :
        self.advance()

    let comentario: string = self.source.substr(start, self.position - 1)
    self.addToken(COMMENT, comentario)

proc handleLabelRef(self: var Tokenizer) =
    let start: int = self.position
    self.advance()

    while self.currentChar.isAlphaNumeric() or self.currentChar == '_':
        self.advance()

    let identifier: string = self.source.substr(start, self.position - 1)
    self.addToken(LABEL_REF, identifier)        

proc tokenize*(self: var Tokenizer): (seq[Token], bool) =
    while self.position < self.source.len:
        case self.currentChar
            of ' ', '\n':
                self.handleWhitespace()
            of '"':
                if (not self.handleString()) :
                    return (newSeq[Token](), false)
            of ',':
                self.addToken(VIRGULA, ",")
                self.advance()
            of '.':
                self.handleDirective()
            of ':':
                self.handleLabelDef()
            of '*':
                self.handleLabelRef()
            of 'a'..'z', 'A'..'Z', '_':
                self.handleIdentifier()
            of '0'..'9', '-':
                self.handleNumber()
            of '$' :
                self.handleRegister()
            of '#' :
                self.handleComment()
            of '(' :
                self.addToken(OPEN_PARENTH, "(")
                self.advance()
            of ')' :
                self.addToken(CLOSE_PARENTH, ")")
                self.advance()
            else:
                echo " - '", "'", self.currentChar
                echo "Character invalido na linha ", self.line
                return (newSeq[Token](), false)
    
    return (self.tokens, true)

File: src/Assembler/tokenizer.go
package assembler

import (
	"fmt"
	"strconv"
)

var KEY_WORDS = map[string]bool{
	"add": true, "addi": true, "addu": true, "addui": true,
	"sub": true, "subi": true, "subu": true, "subui": true,
	"mult": true, "multi": true, "div": true, "divi": true,
	"or": true, "ori": true, "and": true, "andi": true,
	"sll": true, "srl": true,
	"slt": true, "slti": true,
	"li": true, "la": true, "move": true,
	"j": true, "jr": true, "jal": true,
	"beq": true, "bne": true, "bgt": true,
	"bge": true, "blt": true, "ble": true,
	"lw": true, "lb": true,
	"sw": true, "sb": true,
	"lv": true, "sv": true,
	"lrw": true, "lrb": true,
	"inc": true, "dec": true,
	"syscall": true, "return": true, "rand": true,
}

var SECTIONS = map[string]bool{
	".text": true, ".data": true,
}

var TYPES = map[string]bool{
	".int8": true, ".uint8": true,
	".int16": true, ".uint16": true,
	".string": true,
}

var REGISTRERS = map[string]bool{
	"$zero": true, "$t0": true, "$t1": true, "$t2": true, "$t3": true, "$t4": true, "$t5": true,
	"$rt": true, "$gp": true, "$sp": true, "$fp": true, "$sc": true, "$ra": true, "$pc": true, "$ir": true,
}

type TokenType string

const (
	IDENTIFIER   = "IDENTIFIER"
	INSTRUCTION  = "INSTRUCTION"
	REGISTER     = "REGISTER"
	LABEL_DEF    = "LABEL_DEF"
	LABEL_REF    = "LABEL_REF"
	STR_LABEL    = "STR_LABEL"
	INT8_LABEL   = "INT8_LABEL"
	UINT8_LABEL  = "UINT8_LABEL"
	INT16_LABEL  = "INT16_LABEL"
	UINT16_LABEL = "UINT16_LABEL"
	SECTION      = "SECTION"
	TYPE         = "TYPE"
	STRING       = "STRING"
	NUMBER       = "NUMBER"
	VIRGULA      = "VIRGULA"
	NEW_LINE     = "NEW_LINE"
	OPEN_PAREN   = "OPEN_PAREN"
	CLOSE_PAREN  = "CLOSE_PAREN"
)

type Token struct {
	TokenType TokenType
	Value     string
	Line      int
	Column    int
}

func newToken(tt TokenType, value string, line int, column int) *Token {
	t := Token{tt, value, line, column}
	return &t
}

type Tokenizer struct {
	Data     string
	position int
	column   int
	line     int
	len      int
	tokens   []Token
}

func newTokenizer(data string) *Tokenizer {
	tokenizer := Tokenizer{data, 0, 0, 0, 0, []Token{}}
	return &tokenizer
}

func (tokenizer *Tokenizer) addToken(token_type TokenType, value string) {
	token := newToken(token_type, value, tokenizer.line, tokenizer.column)
	tokenizer.tokens = append(tokenizer.tokens, *token)
	tokenizer.len += 1
}

func (tokenizer *Tokenizer) advance() {
	tokenizer.position += 1
	tokenizer.column += 1
}

// func (Tokenizer *Tokenizer) advanceX(value int) {
// 	Tokenizer.position += value
// 	Tokenizer.column += value
// }

func (tokenizer *Tokenizer) nextLine() {
	tokenizer.position += 1
	tokenizer.column = 0
	tokenizer.line += 1
}

func (tokenizer *Tokenizer) getCurrentChar() byte {
	return tokenizer.Data[tokenizer.position]
}

func (tokenizer *Tokenizer) setLastTokenType(tk TokenType) {
	tokenizer.tokens[tokenizer.len-1].TokenType = tk
}

func (tokenizer *Tokenizer) handleSpace() {
	tokenizer.advance()
}

func (tokenizer *Tokenizer) handleNewLine() {
	tokenizer.addToken(NEW_LINE, "\n")
	tokenizer.nextLine()
}

func (tokenizer *Tokenizer) handleVirgula() error {
	last_token := tokenizer.tokens[tokenizer.len-1]

	if last_token.TokenType == REGISTER || last_token.TokenType == NUMBER || last_token.TokenType == LABEL_REF {
		tokenizer.addToken(VIRGULA, ",")
	} else {
		return fmt.Errorf("Virgula servem para separar argumentos em funções. Linha: %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
	}

	tokenizer.advance()
	return nil
}

func (tokenizer *Tokenizer) handleComment() {
	for tokenizer.getCurrentChar() != '\n' {
		tokenizer.advance()
	}
	// tokenizer.nextLine()
}

func isValidCharacterToIdentifier(s byte) bool {
	return (s >= 'a' && s <= 'z') || (s >= 'A' && s <= 'Z') || s == '_' || (s >= '0' && s <= '9')
}

func (tokenizer *Tokenizer) handleIdentifier() error {
	start := tokenizer.position

	current := tokenizer.getCurrentChar()
	for current != ' ' && current != ',' && current != ':' && current != '\n' {
		if !isValidCharacterToIdentifier(current) {
			return fmt.Errorf("Character '%c' invalido para identificador. Linha %d Coluna %d", current, tokenizer.line+1, tokenizer.column+1)
		}
		tokenizer.advance()
		current = tokenizer.getCurrentChar()
	}

	key_world := tokenizer.Data[start:tokenizer.position]
	if tokenizer.getCurrentChar() == ':' {
		if KEY_WORDS[key_world] {
			return fmt.Errorf("Nome invalido para label, '%s' é palavra reservada.", key_world)
		}
		tokenizer.addToken(LABEL_DEF, key_world)
		tokenizer.advance()
	} else if KEY_WORDS[key_world] {
		tokenizer.addToken(INSTRUCTION, key_world)
	} else {
		return fmt.Errorf("'%s' não é um comando valido. Linha %d Coluna %d", key_world, tokenizer.line+1, tokenizer.column+1)
	}
	return nil
}

func (tokenizer *Tokenizer) handleRegister() error {
	start := tokenizer.position
	tokenizer.advance() // Pula o $

	current := tokenizer.getCurrentChar()
	for current != ' ' && current != ',' && current != ')' && current != '(' && current != '\n' {
		if !isValidCharacterToIdentifier(current) {
			return fmt.Errorf("Character '%c' invalido ao referenciar registrador. Linha %d Coluna %d", current, tokenizer.line+1, tokenizer.column+1)
		}
		tokenizer.advance()
		current = tokenizer.getCurrentChar()
	}

	register := tokenizer.Data[start:tokenizer.position]
	if !REGISTRERS[register] {
		return fmt.Errorf("Registrador \"%s\" não é um registrador valido. Linha: %d, Coluna: %d", register, tokenizer.line+1, tokenizer.column+1)
	}

	tokenizer.addToken(REGISTER, register)
	return nil
}

func (tokenizer *Tokenizer) handleLabelRef() error {
	tokenizer.advance() // Pula o *
	start := tokenizer.position

	for tokenizer.getCurrentChar() != ' ' && tokenizer.getCurrentChar() != '\n' {
		if !isValidCharacterToIdentifier(tokenizer.getCurrentChar()) {
			fmt.Printf("%c\n", tokenizer.getCurrentChar())
			return fmt.Errorf("Character invalido ao referenciar label. Linha %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
		}
		tokenizer.advance()
	}

	label := tokenizer.Data[start:tokenizer.position]
	if label == "*" {
		return fmt.Errorf("Definição de referencia a label sem nome. Linha %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
	}

	tokenizer.addToken(LABEL_REF, label)
	// println(label)
	return nil
}

func (tokenizer *Tokenizer) handleSectionOrType() error {
	start := tokenizer.position
	tokenizer.advance() // Pula o '.'

	var current byte
	for {
		current = tokenizer.getCurrentChar()
		if current == ' ' || current == '\n' {
			break
		}

		if !isValidCharacterToIdentifier(current) {
			return fmt.Errorf("Character '%c' invalido ao definir seção ou tipo. Linha %d Coluna %d", current, tokenizer.line+1, tokenizer.column+1)
		}
		tokenizer.advance()
	}

	token := tokenizer.Data[start:tokenizer.position]
	if SECTIONS[token] {
		tokenizer.addToken(SECTION, token)

		var current byte
		for {
			current = tokenizer.getCurrentChar()
			if current == '\n' {
				break
			}

			if current != ' ' {
				return fmt.Errorf("Declaração de seção não recebe argumentos. Linha %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
			}
			tokenizer.advance()
		}

	} else if TYPES[token] {
		switch token {
		case ".string":
			tokenizer.setLastTokenType(STR_LABEL)
		case ".int8":
			tokenizer.setLastTokenType(INT8_LABEL)
		case ".int16":
			tokenizer.setLastTokenType(INT16_LABEL)
		case ".uint8":
			tokenizer.setLastTokenType(UINT8_LABEL)
		case ".uint16":
			tokenizer.setLastTokenType(UINT16_LABEL)
		}
	} else {
		return fmt.Errorf("Nome \"%s\" não é valido para definição de seção ou tipo. Linha %d, Coluna %d", token, tokenizer.line+1, tokenizer.column+1)
	}

	return nil
}

func (tokenizer *Tokenizer) handleNumber() error {
	start := tokenizer.position
	current := tokenizer.getCurrentChar()
	if current == '-' {
		tokenizer.advance()
		current = tokenizer.getCurrentChar()
	}

	if current < '0' || current > '9' {
		if tokenizer.Data[start] == '-' {
			return fmt.Errorf("Dígito esperado após '-'. Linha %d, Coluna %d", tokenizer.line, tokenizer.column)
		}
		return fmt.Errorf("Caractere '%c' não é um dígito. Linha %d, Coluna %d", current, tokenizer.line, tokenizer.column)
	}

	for tokenizer.position < len(tokenizer.Data) {
		current = tokenizer.getCurrentChar()
		if current < '0' || current > '9' {
			break
		}
		tokenizer.advance()
	}
	num := tokenizer.Data[start:tokenizer.position]

	_, err := strconv.Atoi(num)
	if err != nil {
		return fmt.Errorf("Valor invalido ao definir numero. Linha %d, Coluna %d", tokenizer.line+1, tokenizer.column+1)
	}

	tokenizer.addToken(NUMBER, num)
	return nil
}

func (tokenizer *Tokenizer) handleStrings() error {
	tokenizer.advance() // Pula primeira aspas
	start := tokenizer.position

	var last_char byte
	current := tokenizer.getCurrentChar()
	for {
		if current == '\n' {
			return fmt.Errorf("Quebra de linha ao declara string. Linha %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
		}

		tokenizer.advance()
		current = tokenizer.getCurrentChar()
		if current == '"' && last_char != '\\' {
			break
		}

		// println(current, last_char)
		last_char = current

	}

	str := tokenizer.Data[start:tokenizer.position]
	tokenizer.addToken(STRING, str)
	tokenizer.advance()
	return nil
}

func (tokenizer *Tokenizer) handleOpenParenteses() error {
	last_token := tokenizer.tokens[tokenizer.len-1]
	if last_token.TokenType != NUMBER && last_token.TokenType != REGISTER {
		return fmt.Errorf("Mal uso de abertura de parenteses. É esperado offset ou registrador como valor anterior. Linha %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
	}

	tokenizer.advance()
	tokenizer.addToken(OPEN_PAREN, "(")
	return nil
}

func (tokenizer *Tokenizer) handleCloseParenteses() error {
	last_token := tokenizer.tokens[tokenizer.len-1]
	// println(last_token.Value)
	if last_token.TokenType != LABEL_REF && last_token.TokenType != REGISTER {
		return fmt.Errorf("Mal uso de fechamento de parenteses. É esperado label ou registrador como valor anterior. Linha %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
	}

	second_last := tokenizer.tokens[tokenizer.len-2]
	if second_last.TokenType != OPEN_PAREN {
		return fmt.Errorf("Fechamento de parenteses sem abertura. Linha %d Coluna %d", tokenizer.line+1, tokenizer.column+1)
	}

	tokenizer.advance()
	tokenizer.addToken(CLOSE_PAREN, ")")
	return nil
}

func (tokenizer *Tokenizer) Tokenize() error {
	erros := 0
	var err error = nil
	for tokenizer.position != len(tokenizer.Data) {
		current := tokenizer.getCurrentChar()
		if current == ' ' || current == '\t' {
			tokenizer.handleSpace()
		} else if current == '\n' {
			tokenizer.handleNewLine()
		} else if current == '#' {
			tokenizer.handleComment()
		} else if current == ',' {
			err = tokenizer.handleVirgula()
		} else if (current >= 'a' && current <= 'z') || (current >= 'A' && current <= 'Z') || current == '_' {
			err = tokenizer.handleIdentifier()
		} else if current == '$' {
			err = tokenizer.handleRegister()
		} else if current == '*' {
			err = tokenizer.handleLabelRef()
		} else if current == '.' {
			err = tokenizer.handleSectionOrType()
		} else if current == '-' || (current >= '0' && current <= '9') {
			err = tokenizer.handleNumber()
		} else if current == '"' {
			err = tokenizer.handleStrings()
		} else if current == '(' {
			err = tokenizer.handleOpenParenteses()
		} else if current == ')' {
			err = tokenizer.handleCloseParenteses()
		} else {
			return fmt.Errorf("Character invalido na linha %d coluna %d", tokenizer.line+1, tokenizer.column+1)
		}

		if err != nil {
			println(err.Error())
			return fmt.Errorf("%d tokenizer erro(s)", erros)
		}
	}

	return nil
}
